### ISSUE TITLE
Lib should not enforce vertx dependency

### ISSUE THREAD

**Issue by u6f6o** [tsegismont/vertx-contextual-logging#10]
In my opinion, this lib should not enforce a certain vert.x version. When I used this lib the first time, gradle pulled io.vertx.core:4.0.0-CR1 from the repository, while all other vertx dependencies were on CR2 already. 

My temporary workaround was to exclude this transitive dependency: 

```
    implementation ('io.github.tsegismont:vertx-contextual-logging:1.0.0-alpha2') {
        exclude group: 'io.vertx', module: 'vertx-core'
    }
```

**Comment by tsegismont** [comment]
The project depends on Vert.x core for compilation, this is why it's not marked as `optional` in the Maven POM file.

In your Gradle build file, you could make sure the dependency versions are correct for a Vert.x project by adding this platform dependency:

```kotlin
dependencies {
    implementation(platform("io.vertx:vertx-stack-depchain:4.0.0"))

    implementation("io.vertx:vertx-web")
    implementation("io.vertx:vertx-pg-client")
}
```

**Comment by u6f6o** [comment]
I think I (implicitly) did exactly this by using the vertx gradle plugin (https://github.com/jponge/vertx-gradle-plugin) : 
```
plugins {
    id 'org.jetbrains.kotlin.jvm' version '1.4.10'
    id 'io.vertx.vertx-plugin' version '1.1.3'
    id 'org.unbroken-dome.test-sets' version '3.0.1'
    id 'com.google.cloud.tools.jib' version '2.5.0'
}
```
Afaik, this plugin also refers to the vertx-stack-depchain, which kind of dictates the vertx dependencies. It also offers a property to specify the vertx version explicitly: 

```
vertx {
    vertxVersion = "4.0.0.CR2"
}
```
I can use the vertx deps without specifying a version: 

```
dependencies {
    implementation 'io.vertx:vertx-lang-kotlin'
    implementation 'io.vertx:vertx-lang-kotlin-coroutines'
    implementation 'io.vertx:vertx-web'
    implementation 'io.vertx:vertx-circuit-breaker'
    implementation 'io.vertx:vertx-web-client'
    implementation 'io.vertx:vertx-micrometer-metrics'
    implementation 'io.vertx:vertx-config'
    implementation 'io.vertx:vertx-config-yaml'
    implementation 'io.vertx:vertx-web-openapi'

    implementation ('io.github.tsegismont:vertx-contextual-logging:1.0.0-alpha2') {
        exclude group: 'io.vertx', module: 'vertx-core'
    }
```

Yet, when I added the vertx-contextual-logging lib, gradle resolved the vertx core dependency to 4.0.0-CR1 instead of 4.0.0-CR2 which caused a compilation error. 

It's no big deal as I found it rather quickly. Personally I would prefer, that vertx-contextual-logging is either managed by the vertx-stack-depchain or that it uses a specific version of vertx-core only for compile time but ships without it and because of that, requiring the application that uses the library, to provide the right vertx-core version

**Comment by u6f6o** [comment]
Oh, I recognized that contextual-loggign actually pulls milestone2 instead of CR1. To give a bit more context. When I add the dependency, gradle resolves to the following dependencies: 

```
gradle -q dependencies --configuration compileClasspath

------------------------------------------------------------
Root project
------------------------------------------------------------

compileClasspath - Compile classpath for compilation 'main' (target  (jvm)).
+--- javax.xml.bind:jaxb-api:2.3.1
|    \--- javax.activation:javax.activation-api:1.2.0
+--- org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.4.10
|    +--- org.jetbrains.kotlin:kotlin-stdlib:1.4.10
|    |    +--- org.jetbrains.kotlin:kotlin-stdlib-common:1.4.10
|    |    \--- org.jetbrains:annotations:13.0
|    \--- org.jetbrains.kotlin:kotlin-stdlib-jdk7:1.4.10
|         \--- org.jetbrains.kotlin:kotlin-stdlib:1.4.10 (*)
+--- org.jetbrains.kotlin:kotlin-reflect:1.4.10
|    \--- org.jetbrains.kotlin:kotlin-stdlib:1.4.10 (*)
+--- io.vertx:vertx-lang-kotlin -> 4.0.0.CR2
|    +--- io.vertx:vertx-core:4.0.0.CR2 -> 4.0.0-milestone2
|    |    +--- io.netty:netty-common:4.1.39.Final -> 4.1.49.Final
|    |    +--- io.netty:netty-buffer:4.1.39.Final -> 4.1.49.Final
|    |    |    \--- io.netty:netty-common:4.1.49.Final
|    |    +--- io.netty:netty-transport:4.1.39.Final -> 4.1.49.Final
|    |    |    +--- io.netty:netty-common:4.1.49.Final
|    |    |    +--- io.netty:netty-buffer:4.1.49.Final (*)
|    |    |    \--- io.netty:netty-resolver:4.1.49.Final
|    |    |         \--- io.netty:netty-common:4.1.49.Final
|    |    +--- io.netty:netty-handler:4.1.39.Final -> 4.1.49.Final
|    |    |    +--- io.netty:netty-common:4.1.49.Final
|    |    |    +--- io.netty:netty-resolver:4.1.49.Final (*)
|    |    |    +--- io.netty:netty-buffer:4.1.49.Final (*)
|    |    |    +--- io.netty:netty-transport:4.1.49.Final (*)
|    |    |    \--- io.netty:netty-codec:4.1.49.Final
|    |    |         +--- io.netty:netty-common:4.1.49.Final
|    |    |         +--- io.netty:netty-buffer:4.1.49.Final (*)
|    |    |         \--- io.netty:netty-transport:4.1.49.Final (*)
|    |    +--- io.netty:netty-handler-proxy:4.1.39.Final -> 4.1.49.Final
|    |    |    +--- io.netty:netty-common:4.1.49.Final
|    |    |    +--- io.netty:netty-buffer:4.1.49.Final (*)
|    |    |    +--- io.netty:netty-transport:4.1.49.Final (*)
|    |    |    +--- io.netty:netty-codec:4.1.49.Final (*)
|    |    |    +--- io.netty:netty-codec-socks:4.1.49.Final
|    |    |    |    +--- io.netty:netty-common:4.1.49.Final
|    |    |    |    +--- io.netty:netty-buffer:4.1.49.Final (*)
|    |    |    |    +--- io.netty:netty-transport:4.1.49.Final (*)
|    |    |    |    \--- io.netty:netty-codec:4.1.49.Final (*)
|    |    |    \--- io.netty:netty-codec-http:4.1.49.Final
|    |    |         +--- io.netty:netty-common:4.1.49.Final
|    |    |         +--- io.netty:netty-buffer:4.1.49.Final (*)
|    |    |         +--- io.netty:netty-transport:4.1.49.Final (*)
|    |    |         +--- io.netty:netty-codec:4.1.49.Final (*)
|    |    |         \--- io.netty:netty-handler:4.1.49.Final (*)
|    |    +--- io.netty:netty-codec-http:4.1.39.Final -> 4.1.49.Final (*)
|    |    +--- io.netty:netty-codec-http2:4.1.39.Final -> 4.1.49.Final
|    |    |    +--- io.netty:netty-common:4.1.49.Final
|    |    |    +--- io.netty:netty-buffer:4.1.49.Final (*)
|    |    |    +--- io.netty:netty-transport:4.1.49.Final (*)
|    |    |    +--- io.netty:netty-codec:4.1.49.Final (*)
|    |    |    +--- io.netty:netty-handler:4.1.49.Final (*)
|    |    |    \--- io.netty:netty-codec-http:4.1.49.Final (*)
|    |    +--- io.netty:netty-resolver:4.1.39.Final -> 4.1.49.Final (*)
|    |    +--- io.netty:netty-resolver-dns:4.1.39.Final -> 4.1.49.Final
|    |    |    +--- io.netty:netty-common:4.1.49.Final
|    |    |    +--- io.netty:netty-buffer:4.1.49.Final (*)
|    |    |    +--- io.netty:netty-resolver:4.1.49.Final (*)
|    |    |    +--- io.netty:netty-transport:4.1.49.Final (*)
|    |    |    +--- io.netty:netty-codec:4.1.49.Final (*)
|    |    |    \--- io.netty:netty-codec-dns:4.1.49.Final
|    |    |         +--- io.netty:netty-common:4.1.49.Final
|    |    |         +--- io.netty:netty-buffer:4.1.49.Final (*)
|    |    |         +--- io.netty:netty-transport:4.1.49.Final (*)
|    |    |         \--- io.netty:netty-codec:4.1.49.Final (*)
|    |    +--- com.fasterxml.jackson.core:jackson-core:2.9.9 -> 2.11.3
|    |    \--- com.fasterxml.jackson.core:jackson-databind:2.9.9.1 -> 2.11.3
|    |         +--- com.fasterxml.jackson.core:jackson-annotations:2.11.3
|    |         \--- com.fasterxml.jackson.core:jackson-core:2.11.3
|    \--- org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.4.10 (*)
+--- io.vertx:vertx-lang-kotlin-coroutines -> 4.0.0.CR2
|    +--- org.jetbrains.kotlinx:kotlinx-coroutines-core:1.3.9
|    |    \--- org.jetbrains.kotlinx:kotlinx-coroutines-core-jvm:1.3.9
|    +--- org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.4.10 (*)
|    \--- io.vertx:vertx-core:4.0.0.CR2 -> 4.0.0-milestone2 (*)
+--- io.vertx:vertx-web -> 4.0.0.CR2
|    +--- io.vertx:vertx-web-common:4.0.0.CR2
|    |    \--- io.vertx:vertx-core:4.0.0.CR2 -> 4.0.0-milestone2 (*)
|    +--- io.vertx:vertx-auth-common:4.0.0.CR2
|    |    \--- io.vertx:vertx-core:4.0.0.CR2 -> 4.0.0-milestone2 (*)
|    +--- io.vertx:vertx-bridge-common:4.0.0.CR2
|    \--- io.vertx:vertx-core:4.0.0.CR2 -> 4.0.0-milestone2 (*)
+--- io.vertx:vertx-circuit-breaker -> 4.0.0.CR2
|    +--- org.hdrhistogram:HdrHistogram:2.1.10 -> 2.1.12
|    \--- io.vertx:vertx-core:4.0.0.CR2 -> 4.0.0-milestone2 (*)
+--- io.vertx:vertx-web-client -> 4.0.0.CR2
|    +--- io.vertx:vertx-web-common:4.0.0.CR2 (*)
|    +--- io.vertx:vertx-auth-common:4.0.0.CR2 (*)
|    \--- io.vertx:vertx-core:4.0.0.CR2 -> 4.0.0-milestone2 (*)
+--- io.vertx:vertx-micrometer-metrics -> 4.0.0.CR2
|    +--- io.micrometer:micrometer-core:1.5.2 -> 1.5.5
|    |    \--- org.hdrhistogram:HdrHistogram:2.1.12
|    +--- org.hdrhistogram:HdrHistogram:2.1.10 -> 2.1.12
|    \--- io.vertx:vertx-core:4.0.0.CR2 -> 4.0.0-milestone2 (*)
+--- io.vertx:vertx-config -> 4.0.0.CR2
|    \--- io.vertx:vertx-core:4.0.0.CR2 -> 4.0.0-milestone2 (*)
+--- io.vertx:vertx-config-yaml -> 4.0.0.CR2
|    +--- io.vertx:vertx-config:4.0.0.CR2 (*)
|    +--- com.fasterxml.jackson.core:jackson-databind:2.11.3 (*)
|    +--- com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:2.11.3
|    |    +--- com.fasterxml.jackson.core:jackson-databind:2.11.3 (*)
|    |    +--- org.yaml:snakeyaml:1.26
|    |    \--- com.fasterxml.jackson.core:jackson-core:2.11.3
|    \--- io.vertx:vertx-core:4.0.0.CR2 -> 4.0.0-milestone2 (*)
+--- io.vertx:vertx-web-openapi -> 4.0.0.CR2
|    +--- io.vertx:vertx-web:4.0.0.CR2 (*)
|    +--- io.vertx:vertx-web-validation:4.0.0.CR2
|    |    +--- io.vertx:vertx-web:4.0.0.CR2 (*)
|    |    +--- io.vertx:vertx-json-schema:4.0.0.CR2
|    |    |    \--- io.vertx:vertx-core:4.0.0.CR2 -> 4.0.0-milestone2 (*)
|    |    \--- io.vertx:vertx-core:4.0.0.CR2 -> 4.0.0-milestone2 (*)
|    +--- io.vertx:vertx-web-api-service:4.0.0.CR2
|    |    +--- io.vertx:vertx-service-proxy:4.0.0.CR2
|    |    |    \--- io.vertx:vertx-core:4.0.0.CR2 -> 4.0.0-milestone2 (*)
|    |    +--- io.vertx:vertx-web-validation:4.0.0.CR2 (*)
|    |    \--- io.vertx:vertx-core:4.0.0.CR2 -> 4.0.0-milestone2 (*)
|    +--- io.vertx:vertx-json-schema:4.0.0.CR2 (*)
|    +--- com.fasterxml.jackson.core:jackson-databind:2.11.3 (*)
|    +--- com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:2.11.3 (*)
|    \--- io.vertx:vertx-core:4.0.0.CR2 -> 4.0.0-milestone2 (*)
+--- io.github.tsegismont:vertx-contextual-logging:1.0.0-alpha2
|    \--- io.vertx:vertx-core:4.0.0-milestone2 (*)
+--- ch.qos.logback:logback-classic:1.2.3
|    +--- ch.qos.logback:logback-core:1.2.3
|    \--- org.slf4j:slf4j-api:1.7.25 -> 1.7.29
+--- io.sentry:sentry-logback:1.7.30
|    +--- io.sentry:sentry:1.7.30
|    |    +--- org.slf4j:slf4j-api:1.7.24 -> 1.7.29
|    |    \--- com.fasterxml.jackson.core:jackson-core:2.10.0 -> 2.11.3
|    +--- org.slf4j:slf4j-api:1.7.24 -> 1.7.29
|    +--- ch.qos.logback:logback-core:1.2.1 -> 1.2.3
|    \--- ch.qos.logback:logback-classic:1.2.1 -> 1.2.3 (*)
+--- io.github.microutils:kotlin-logging:2.0.3
|    \--- io.github.microutils:kotlin-logging-jvm:2.0.3
|         +--- org.slf4j:slf4j-api:1.7.29
|         +--- org.jetbrains.kotlin:kotlin-stdlib:1.4.10 (*)
|         \--- org.jetbrains.kotlin:kotlin-stdlib-common:1.4.10
+--- com.fasterxml.jackson.module:jackson-module-kotlin:2.11.2
|    +--- com.fasterxml.jackson.core:jackson-databind:2.11.2 -> 2.11.3 (*)
|    +--- com.fasterxml.jackson.core:jackson-annotations:2.11.2 -> 2.11.3
|    \--- org.jetbrains.kotlin:kotlin-reflect:1.3.72 -> 1.4.10 (*)
+--- io.jsonwebtoken:jjwt-api:0.11.2
+--- io.vertx:vertx-stack-depchain:4.0.0.CR2
|    +--- io.vertx:vertx-core:4.0.0.CR2 -> 4.0.0-milestone2 (c)
|    +--- io.vertx:vertx-lang-kotlin:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-lang-kotlin-coroutines:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-micrometer-metrics:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-web:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-web-client:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-web-openapi:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-circuit-breaker:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-config:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-config-yaml:4.0.0.CR2 (c)
|    +--- com.fasterxml.jackson.core:jackson-databind:2.11.3 (c)
|    +--- com.fasterxml.jackson.core:jackson-annotations:2.11.3 (c)
|    +--- io.vertx:vertx-web-common:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-auth-common:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-bridge-common:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-web-validation:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-web-api-service:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-json-schema:4.0.0.CR2 (c)
|    +--- com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:2.11.3 (c)
|    +--- io.netty:netty-common:4.1.49.Final (c)
|    +--- io.netty:netty-buffer:4.1.49.Final (c)
|    +--- io.netty:netty-transport:4.1.49.Final (c)
|    +--- io.netty:netty-handler:4.1.49.Final (c)
|    +--- io.netty:netty-handler-proxy:4.1.49.Final (c)
|    +--- io.netty:netty-codec-http:4.1.49.Final (c)
|    +--- io.netty:netty-codec-http2:4.1.49.Final (c)
|    +--- io.netty:netty-resolver:4.1.49.Final (c)
|    +--- io.netty:netty-resolver-dns:4.1.49.Final (c)
|    +--- com.fasterxml.jackson.core:jackson-core:2.11.3 (c)
|    \--- io.vertx:vertx-service-proxy:4.0.0.CR2 (c)
\--- io.vertx:vertx-core -> 4.0.0-milestone2 (*)

(c) - dependency constraint
(*) - dependencies omitted (listed previously)

A web-based, searchable dependency report is available by adding the --scan option.
```
Basically, most of the vertx-deps are resolved to milestone2 instead of cr2. When I remove it as a dependency, all vert.x deps are resolved to CR2: 

```
gradle -q dependencies --configuration compileClasspath

------------------------------------------------------------
Root project
------------------------------------------------------------

compileClasspath - Compile classpath for compilation 'main' (target  (jvm)).
+--- javax.xml.bind:jaxb-api:2.3.1
|    \--- javax.activation:javax.activation-api:1.2.0
+--- org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.4.10
|    +--- org.jetbrains.kotlin:kotlin-stdlib:1.4.10
|    |    +--- org.jetbrains.kotlin:kotlin-stdlib-common:1.4.10
|    |    \--- org.jetbrains:annotations:13.0
|    \--- org.jetbrains.kotlin:kotlin-stdlib-jdk7:1.4.10
|         \--- org.jetbrains.kotlin:kotlin-stdlib:1.4.10 (*)
+--- org.jetbrains.kotlin:kotlin-reflect:1.4.10
|    \--- org.jetbrains.kotlin:kotlin-stdlib:1.4.10 (*)
+--- io.vertx:vertx-lang-kotlin -> 4.0.0.CR2
|    +--- io.vertx:vertx-core:4.0.0.CR2
|    |    +--- io.netty:netty-common:4.1.49.Final
|    |    +--- io.netty:netty-buffer:4.1.49.Final
|    |    |    \--- io.netty:netty-common:4.1.49.Final
|    |    +--- io.netty:netty-transport:4.1.49.Final
|    |    |    +--- io.netty:netty-common:4.1.49.Final
|    |    |    +--- io.netty:netty-buffer:4.1.49.Final (*)
|    |    |    \--- io.netty:netty-resolver:4.1.49.Final
|    |    |         \--- io.netty:netty-common:4.1.49.Final
|    |    +--- io.netty:netty-handler:4.1.49.Final
|    |    |    +--- io.netty:netty-common:4.1.49.Final
|    |    |    +--- io.netty:netty-resolver:4.1.49.Final (*)
|    |    |    +--- io.netty:netty-buffer:4.1.49.Final (*)
|    |    |    +--- io.netty:netty-transport:4.1.49.Final (*)
|    |    |    \--- io.netty:netty-codec:4.1.49.Final
|    |    |         +--- io.netty:netty-common:4.1.49.Final
|    |    |         +--- io.netty:netty-buffer:4.1.49.Final (*)
|    |    |         \--- io.netty:netty-transport:4.1.49.Final (*)
|    |    +--- io.netty:netty-handler-proxy:4.1.49.Final
|    |    |    +--- io.netty:netty-common:4.1.49.Final
|    |    |    +--- io.netty:netty-buffer:4.1.49.Final (*)
|    |    |    +--- io.netty:netty-transport:4.1.49.Final (*)
|    |    |    +--- io.netty:netty-codec:4.1.49.Final (*)
|    |    |    +--- io.netty:netty-codec-socks:4.1.49.Final
|    |    |    |    +--- io.netty:netty-common:4.1.49.Final
|    |    |    |    +--- io.netty:netty-buffer:4.1.49.Final (*)
|    |    |    |    +--- io.netty:netty-transport:4.1.49.Final (*)
|    |    |    |    \--- io.netty:netty-codec:4.1.49.Final (*)
|    |    |    \--- io.netty:netty-codec-http:4.1.49.Final
|    |    |         +--- io.netty:netty-common:4.1.49.Final
|    |    |         +--- io.netty:netty-buffer:4.1.49.Final (*)
|    |    |         +--- io.netty:netty-transport:4.1.49.Final (*)
|    |    |         +--- io.netty:netty-codec:4.1.49.Final (*)
|    |    |         \--- io.netty:netty-handler:4.1.49.Final (*)
|    |    +--- io.netty:netty-codec-http:4.1.49.Final (*)
|    |    +--- io.netty:netty-codec-http2:4.1.49.Final
|    |    |    +--- io.netty:netty-common:4.1.49.Final
|    |    |    +--- io.netty:netty-buffer:4.1.49.Final (*)
|    |    |    +--- io.netty:netty-transport:4.1.49.Final (*)
|    |    |    +--- io.netty:netty-codec:4.1.49.Final (*)
|    |    |    +--- io.netty:netty-handler:4.1.49.Final (*)
|    |    |    \--- io.netty:netty-codec-http:4.1.49.Final (*)
|    |    +--- io.netty:netty-resolver:4.1.49.Final (*)
|    |    +--- io.netty:netty-resolver-dns:4.1.49.Final
|    |    |    +--- io.netty:netty-common:4.1.49.Final
|    |    |    +--- io.netty:netty-buffer:4.1.49.Final (*)
|    |    |    +--- io.netty:netty-resolver:4.1.49.Final (*)
|    |    |    +--- io.netty:netty-transport:4.1.49.Final (*)
|    |    |    +--- io.netty:netty-codec:4.1.49.Final (*)
|    |    |    \--- io.netty:netty-codec-dns:4.1.49.Final
|    |    |         +--- io.netty:netty-common:4.1.49.Final
|    |    |         +--- io.netty:netty-buffer:4.1.49.Final (*)
|    |    |         +--- io.netty:netty-transport:4.1.49.Final (*)
|    |    |         \--- io.netty:netty-codec:4.1.49.Final (*)
|    |    \--- com.fasterxml.jackson.core:jackson-core:2.11.3
|    \--- org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.4.10 (*)
+--- io.vertx:vertx-lang-kotlin-coroutines -> 4.0.0.CR2
|    +--- org.jetbrains.kotlinx:kotlinx-coroutines-core:1.3.9
|    |    \--- org.jetbrains.kotlinx:kotlinx-coroutines-core-jvm:1.3.9
|    +--- org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.4.10 (*)
|    \--- io.vertx:vertx-core:4.0.0.CR2 (*)
+--- io.vertx:vertx-web -> 4.0.0.CR2
|    +--- io.vertx:vertx-web-common:4.0.0.CR2
|    |    \--- io.vertx:vertx-core:4.0.0.CR2 (*)
|    +--- io.vertx:vertx-auth-common:4.0.0.CR2
|    |    \--- io.vertx:vertx-core:4.0.0.CR2 (*)
|    +--- io.vertx:vertx-bridge-common:4.0.0.CR2
|    \--- io.vertx:vertx-core:4.0.0.CR2 (*)
+--- io.vertx:vertx-circuit-breaker -> 4.0.0.CR2
|    +--- org.hdrhistogram:HdrHistogram:2.1.10 -> 2.1.12
|    \--- io.vertx:vertx-core:4.0.0.CR2 (*)
+--- io.vertx:vertx-web-client -> 4.0.0.CR2
|    +--- io.vertx:vertx-web-common:4.0.0.CR2 (*)
|    +--- io.vertx:vertx-auth-common:4.0.0.CR2 (*)
|    \--- io.vertx:vertx-core:4.0.0.CR2 (*)
+--- io.vertx:vertx-micrometer-metrics -> 4.0.0.CR2
|    +--- io.micrometer:micrometer-core:1.5.2 -> 1.5.5
|    |    \--- org.hdrhistogram:HdrHistogram:2.1.12
|    +--- org.hdrhistogram:HdrHistogram:2.1.10 -> 2.1.12
|    \--- io.vertx:vertx-core:4.0.0.CR2 (*)
+--- io.vertx:vertx-config -> 4.0.0.CR2
|    \--- io.vertx:vertx-core:4.0.0.CR2 (*)
+--- io.vertx:vertx-config-yaml -> 4.0.0.CR2
|    +--- io.vertx:vertx-config:4.0.0.CR2 (*)
|    +--- com.fasterxml.jackson.core:jackson-databind:2.11.3
|    |    +--- com.fasterxml.jackson.core:jackson-annotations:2.11.3
|    |    \--- com.fasterxml.jackson.core:jackson-core:2.11.3
|    +--- com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:2.11.3
|    |    +--- com.fasterxml.jackson.core:jackson-databind:2.11.3 (*)
|    |    +--- org.yaml:snakeyaml:1.26
|    |    \--- com.fasterxml.jackson.core:jackson-core:2.11.3
|    \--- io.vertx:vertx-core:4.0.0.CR2 (*)
+--- io.vertx:vertx-web-openapi -> 4.0.0.CR2
|    +--- io.vertx:vertx-web:4.0.0.CR2 (*)
|    +--- io.vertx:vertx-web-validation:4.0.0.CR2
|    |    +--- io.vertx:vertx-web:4.0.0.CR2 (*)
|    |    +--- io.vertx:vertx-json-schema:4.0.0.CR2
|    |    |    \--- io.vertx:vertx-core:4.0.0.CR2 (*)
|    |    \--- io.vertx:vertx-core:4.0.0.CR2 (*)
|    +--- io.vertx:vertx-web-api-service:4.0.0.CR2
|    |    +--- io.vertx:vertx-service-proxy:4.0.0.CR2
|    |    |    \--- io.vertx:vertx-core:4.0.0.CR2 (*)
|    |    +--- io.vertx:vertx-web-validation:4.0.0.CR2 (*)
|    |    \--- io.vertx:vertx-core:4.0.0.CR2 (*)
|    +--- io.vertx:vertx-json-schema:4.0.0.CR2 (*)
|    +--- com.fasterxml.jackson.core:jackson-databind:2.11.3 (*)
|    +--- com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:2.11.3 (*)
|    \--- io.vertx:vertx-core:4.0.0.CR2 (*)
+--- ch.qos.logback:logback-classic:1.2.3
|    +--- ch.qos.logback:logback-core:1.2.3
|    \--- org.slf4j:slf4j-api:1.7.25 -> 1.7.29
+--- io.sentry:sentry-logback:1.7.30
|    +--- io.sentry:sentry:1.7.30
|    |    +--- org.slf4j:slf4j-api:1.7.24 -> 1.7.29
|    |    \--- com.fasterxml.jackson.core:jackson-core:2.10.0 -> 2.11.3
|    +--- org.slf4j:slf4j-api:1.7.24 -> 1.7.29
|    +--- ch.qos.logback:logback-core:1.2.1 -> 1.2.3
|    \--- ch.qos.logback:logback-classic:1.2.1 -> 1.2.3 (*)
+--- io.github.microutils:kotlin-logging:2.0.3
|    \--- io.github.microutils:kotlin-logging-jvm:2.0.3
|         +--- org.slf4j:slf4j-api:1.7.29
|         +--- org.jetbrains.kotlin:kotlin-stdlib:1.4.10 (*)
|         \--- org.jetbrains.kotlin:kotlin-stdlib-common:1.4.10
+--- com.fasterxml.jackson.module:jackson-module-kotlin:2.11.2
|    +--- com.fasterxml.jackson.core:jackson-databind:2.11.2 -> 2.11.3 (*)
|    +--- com.fasterxml.jackson.core:jackson-annotations:2.11.2 -> 2.11.3
|    \--- org.jetbrains.kotlin:kotlin-reflect:1.3.72 -> 1.4.10 (*)
+--- io.jsonwebtoken:jjwt-api:0.11.2
+--- io.vertx:vertx-stack-depchain:4.0.0.CR2
|    +--- io.vertx:vertx-core:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-lang-kotlin:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-lang-kotlin-coroutines:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-micrometer-metrics:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-web:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-web-client:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-web-openapi:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-circuit-breaker:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-config:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-config-yaml:4.0.0.CR2 (c)
|    +--- com.fasterxml.jackson.core:jackson-databind:2.11.3 (c)
|    +--- com.fasterxml.jackson.core:jackson-annotations:2.11.3 (c)
|    +--- io.netty:netty-common:4.1.49.Final (c)
|    +--- io.netty:netty-buffer:4.1.49.Final (c)
|    +--- io.netty:netty-transport:4.1.49.Final (c)
|    +--- io.netty:netty-handler:4.1.49.Final (c)
|    +--- io.netty:netty-handler-proxy:4.1.49.Final (c)
|    +--- io.netty:netty-codec-http:4.1.49.Final (c)
|    +--- io.netty:netty-codec-http2:4.1.49.Final (c)
|    +--- io.netty:netty-resolver:4.1.49.Final (c)
|    +--- io.netty:netty-resolver-dns:4.1.49.Final (c)
|    +--- com.fasterxml.jackson.core:jackson-core:2.11.3 (c)
|    +--- io.vertx:vertx-web-common:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-auth-common:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-bridge-common:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-web-validation:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-web-api-service:4.0.0.CR2 (c)
|    +--- io.vertx:vertx-json-schema:4.0.0.CR2 (c)
|    +--- com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:2.11.3 (c)
|    \--- io.vertx:vertx-service-proxy:4.0.0.CR2 (c)
\--- io.vertx:vertx-core -> 4.0.0.CR2 (*)

(c) - dependency constraint
(*) - dependencies omitted (listed previously)

A web-based, searchable dependency report is available by adding the --scan option.
```

**Comment by tsegismont** [comment]
@jponge does the Vert.x Gradle plugin use platform dependencies?

@u6f6o according to [Importing Maven BOMs](https://docs.gradle.org/current/userguide/platforms.html#sub:bom_import), you could use `enforcedPlatform` to make sure versions defined in `vertx-stack-depchain` override any other version found in the dependency graph:

```kotlin
dependencies {
    implementation(enforcedPlatform("io.vertx:vertx-stack-depchain:4.0.0"))

    implementation("io.vertx:vertx-web")
    implementation("io.vertx:vertx-pg-client")
}
```

**Comment by jponge** [comment]
Right now the plugin does:

```kotlin
project.dependencies.apply {
  add("implementation", platform("io.vertx:vertx-stack-depchain:$vertxVersion"))
}
```

**Comment by jponge** [comment]
Sounds like `enforcedPlatform` would be better indeed

**Comment by u6f6o** [comment]
I tried it with `enforcedPlatform` and every vertx-dependency remained on 4.0.0 instead of 4.0.0.milestone2 as before.

**Comment by tsegismont** [comment]
@u6f6o Looking at Gradle [version conflict resolution](https://docs.gradle.org/current/userguide/dependency_resolution.html#sec:version-conflict) doc, I believe the issue is related to semantic versioning.

If you don't use `enforcedPlatform`, Gradle uses its default resolution system which consists in selecting the highest version. If version format follows semantic versioning guidelines, it should work, but `4.0.0-milestone2` doesn't. It must be seen as higher than `4.0.0` or `4.0.0.CR2`.

Now that `vertx-contextual-logging` has an updated version (`4.0.0`, thanks to your PR #9), this shouldn't happen again.

**Comment by tsegismont** [comment]
Closing as fixed by #9
